mirrord/cli/src/port_forward.rs

```diff
@@ -442,12 +420,255 @@ impl PortForwarder {
+                    self.agent_connection.sender.send(ClientMessage::Ping).await?;
+                    self.waiting_for_pong = true;
+                    self.ping_pong_timeout = Instant::now() + Duration::from_secs(30);
+                },
+
+                message = self.agent_connection.receiver.recv() => match message {
+                    Some(message) => {
+                        self.handle_msg_from_agent(message).await?},
+                    None => {
+                        break Err(PortForwardError::AgentError("unexpected end of connection with agent".into()));
+                    },
+                },
+
+                Some((task_id, update)) = self.background_tasks.next() => {
+                    // dbg!((&task_id, &update));
+                    self.handle_msg_from_local(task_id, update).await?
+                },
+            }
+        }
+    }
+
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    async fn handle_msg_from_agent(
+        &mut self,
+        message: DaemonMessage,
+    ) -> Result<(), PortForwardError> {
+        match message {
+            DaemonMessage::Tcp(msg) => {
+                self.incoming_proxy
+                    .send(IncomingProxyMessage::AgentMirror(msg))
+                    .await
+            }
+            DaemonMessage::TcpSteal(msg) => {
+                self.incoming_proxy
+                    .send(IncomingProxyMessage::AgentSteal(msg))
+                    .await
+            }
+            DaemonMessage::LogMessage(log_message) => match log_message.level {
+                LogLevel::Warn => tracing::warn!("agent log: {}", log_message.message),
+                LogLevel::Error => tracing::error!("agent log: {}", log_message.message),
+            },
+            DaemonMessage::Close(error) => {
+                return Err(PortForwardError::AgentError(error));
+            }
+            DaemonMessage::Pong if self.waiting_for_pong => {
+                self.waiting_for_pong = false;
+            }
+            other => {
+                // includes unexepcted DaemonMessage::Pong
+                return Err(PortForwardError::AgentError(format!(
+                    "unexpected message from agent: {other:?}"
+                )));
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = Level::TRACE, skip(self), err)]
+    async fn handle_msg_from_local(
+        &mut self,
+        task_id: MainTaskId,
+        update: TaskUpdate<ProxyMessage, IntProxyError>,
+    ) -> Result<(), PortForwardError> {
+        match (task_id, update) {
+            (MainTaskId::IncomingProxy, TaskUpdate::Message(message)) => match message {
+                ProxyMessage::ToAgent(message) => {
+                    if matches!(
+                        message,
+                        ClientMessage::TcpSteal(LayerTcpSteal::PortSubscribe(_))
+                    ) || matches!(message, ClientMessage::Tcp(LayerTcp::PortSubscribe(_)))
```

Comment:
I think this could be shortened a bit with something like this, but up to you.

```suggestion
                    if matches!(
                        message,
                        ClientMessage::TcpSteal(LayerTcpSteal::PortSubscribe(_))
                        | ClientMessage::Tcp(LayerTcp::PortSubscribe(_)))
```